src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+    IntColumnBuilder streamSizeCol = new IntColumnBuilder();
+    IntColumnBuilder sampCountCol = new IntColumnBuilder();
+    StringColumnBuilder sampTypeCol = new StringColumnBuilder();
+    for (int i = 0; i < streamCounts.size(); i++) {
+      wIdCol.appendInt(i + 1);
+      streamSizeCol.appendInt(streamCounts.get(i));
+      sampCountCol.appendInt(sampleCounts[i]);
+      sampTypeCol.appendString(sampleType.name());
+    }
+    ImmutableList.Builder<Column<?>> columns = ImmutableList.builder();
+    columns.add(wIdCol.build(), streamSizeCol.build(), sampCountCol.build(),
+        sampTypeCol.build());
+    setEOS();
+    return new TupleBatch(SCHEMA, columns.build());
+  }
+
+  /**
+   * Creates a WithReplacement distribution across the workers.
+   * 
+   * @param tupleCounts
+   *          list of how many tuples each worker has.
+   * @param sampleSize
+   *          total number of samples to distribute across the workers.
+   * @return array representing the distribution across the workers.
+   */
+  private int[] withReplacementDistribution(List<Integer> tupleCounts,
+      int sampleSize) {
+    int[] distribution = new int[tupleCounts.size()];
+    int totalTupleCount = 0;
+    for (int val : tupleCounts) {
+      totalTupleCount += val;
+    }
+
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (int j = 0; j < tupleCounts.size(); j++) {
+        if (sampleTupleIdx < tupleCounts.get(j) + tupleOffset) {
+          distribution[j] += 1;
+          break;
+        }
+        tupleOffset += tupleCounts.get(j);
+      }
+    }
+    return distribution;
+  }
+
+  /**
+   * Creates a WithoutReplacement distribution across the workers.
+   *
+   * @param tupleCounts
+   *          list of how many tuples each worker has.
+   * @param sampleSize
+   *          total number of samples to distribute across the workers.
+   * @return array representing the distribution across the workers.
+   */
+  private int[] withoutReplacementDistribution(List<Integer> tupleCounts,
+      int sampleSize) {
+    int[] distribution = new int[tupleCounts.size()];
+    int totalTupleCount = 0;
+    for (int val : tupleCounts) {
+      totalTupleCount += val;
+    }
+    List<Integer> logicalTupleCounts = new ArrayList<>(tupleCounts);
+
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount - i);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (int j = 0; j < logicalTupleCounts.size(); j++) {
+        if (sampleTupleIdx < logicalTupleCounts.get(j) + tupleOffset) {
+          distribution[j] += 1;
+          // Cannot sample the same tuple, so pretend it doesn't exist anymore.
+          logicalTupleCounts.set(j, logicalTupleCounts.get(j) - 1);
```

Comment:
I can describe the detailed decision here in person sometime, but the main idea is that it's simulating the removal of the tuple from the population. I don't have a formal proof that this is legitimate to do, but it does pass all the tests I've thrown at it :)